lib/services/local/access_list.go

```diff
@@ -303,6 +320,84 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
 	return member, trace.Wrap(err)
 }
 
+// GetAccessListOwners returns a list of all owners in an Access List, including those inherited from nested Access Lists.
+//
+// Returned Owners are not validated for ownership requirements – use `IsAccessListOwner` for validation.
+func (a *AccessListService) GetAccessListOwners(ctx context.Context, accessListName string) ([]*accesslist.Owner, error) {
+	seen := map[string]struct{}{}
+	type queueItem struct {
+		name  string
+		depth int
+	}
+	queue := []queueItem{{name: accessListName, depth: 0}}
+	var allOwners []*accesslist.Owner
+
+	for len(queue) > 0 {
+		item := queue[0]
+		queue = queue[1:]
+
+		if item.depth > accesslist.MaxAllowedDepth {
+			a.log.Warnf("Exceeded maximum depth of %d while gathering nested owners for access list %s", accesslist.MaxAllowedDepth, accessListName)
+			continue
+		}
+
+		accessList, err := a.GetAccessList(ctx, item.name)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Owners are inherited from members of nested lists.
+		// So, on l0, we want to grab all the owners from the provided accessListName,
+		// while > l0, we want to fetch members of the nested lists.
+		if item.name == accessListName {
+			for _, owner := range accessList.GetOwners() {
+				if owner.MembershipKind == accesslist.MembershipKindList {
+					if _, ok := seen[owner.Name]; ok {
+						continue
+					}
+					seen[owner.Name] = struct{}{}
+					queue = append(queue, queueItem{name: owner.Name, depth: item.depth + 1})
+				} else {
+					allOwners = append(allOwners, &owner)
+				}
+			}
+		} else {
+			var pageToken string
+			var members []*accesslist.AccessListMember
+
+			for {
+				fetchedMembers, pageToken, err := a.ListAccessListMembers(ctx, item.name, 0, pageToken)
```

Comment:
It looks like the page token logic is broken here 

note the on each following call:
```
fetchedMembers, pageToken, err := a.ListAccessListMembers(ctx, item.name, 0, pageToken)
```
the pageToken is scoped to for loop and reacted on each `ListAccessListMembers` call where the top 
```
var pageToken string
```

is never updated so `pageToken` injected to   `a.ListAccessListMembers(ctx, item.name, 0, pageToken)` always contains  `""`

in the result this logic creates a spin loop that existing backend resources.

Reply:
@kiosion This is critical to fix, this class of issues with pagination has caused production outages and aborted deployments in Cloud before.

Also, can this be factored out into a helper method that returns all members with proper pagination? This function is long as-is.

Reply:
The scoping issue w/ `pageToken` should be fixed.

Which parts would you want split into a helper? The following, which just fetches the paginated list of all members, or the whole second half of the func?

https://github.com/gravitational/teleport/blob/da585af4e9a2970ceedf9b5d8c9acd000e50f0c8/lib/services/local/access_list.go#L367-L380

Reply:
Yeah I meant the helper for fetching all list members. We already have some example, even for access lists actually, maybe we can just export and reuse it? https://github.com/gravitational/teleport.e/blob/adb418b4ff297f036d3024116f38b1eb7f818121/lib/accesslist/reconciler_ineligible_status.go#L205-L220